cloud/openstack/os_keystone_endpoint.py

```diff
@@ -0,0 +1,334 @@
+    """
+    if interface:
+        endpoints_list = cloud.search_endpoints(
+            filters={'service_id': service, 'interface': interface})
+    else:
+        endpoints_list = cloud.search_endpoints(
+            filters={'service_id': service})
+    if len(endpoints_list) > 1:
+        module.fail_json(msg='Multiple endpoint entries exist service %s.'
+                             % service)
+    return endpoints_list[0] if len(endpoints_list) == 1 else None
+
+
+def main():
+
+    global cloud, module
+
+    argument_spec = openstack_full_argument_spec(
+        service_name=dict(required=True),
+        service_type=dict(required=True),
+        interface=dict(required=False),
+        url=dict(required=False),
+        admin_url=dict(required=False),
+        public_url=dict(required=False),
+        internal_url=dict(required=False),
+        region=dict(required=False),
+        enabled=dict(default=True, required=False),
+        state=dict(default='present', choices=['absent', 'present'])
+    )
+
+    module_kwargs = openstack_module_kwargs(
+        required_one_of=[
+            ['url', 'admin_url', 'internal_url', 'public_url']
+        ],
+        mutually_exclusive=[
+            ['url', 'admin_url'],
+            ['url', 'internal_url'],
+            ['url', 'public_url'],
+        ],
+        required_together=[['interface', 'url']]
+    )
+
+    module = AnsibleModule(argument_spec,
+                           supports_check_mode=True,
+                           **module_kwargs)
+    # role grant/revoke API introduced in 1.11.0
+    if not HAS_SHADE or \
+            (StrictVersion(shade.__version__) < StrictVersion('1.11.0')):
+        module.fail_json(msg='shade 1.11.0 or higher is required for this module')
+
+    try:
+
+        cloud = shade.operator_cloud(**module.params)
+
+        service_name = module.params.get('service_name', None)
+        service_type = module.params.get('service_type', None)
+        interface = module.params.get('interface', None)
+        url = module.params.get('url', None)
+        region = module.params.get('region', None)
+        enabled = "True" == module.params.get('enabled', True)
+        state = module.params.get('state', None)
+        admin_url = module.params.get('admin_url', None)
+        internal_url = module.params.get('internal_url', None)
+        public_url = module.params.get('public_url', None)
+
+        services_list = cloud.search_services(
+            filters={'name': service_name, 'type': service_type})
+        if len(services_list) != 1:
+            module.fail_json(
+               msg='Service %s not found or more than 1 service have this name'
+                    % service_name)
+        # Only service id is capable to identify unique entry.
+        service_id = services_list[0].id
+
+        current_endpoint = get_endpoint(service_id, interface)
+
+        if current_endpoint is None and state == 'absent':
+            module.exit_json(changed=False)
```

Comment:
See the functions named '_system_state_change' and '_needs_update' in the other ansible shade modules and how those are structured. Please implement that here (rather than in the main() function)